File: cpg-library/src/test/java/de/fraunhofer/aisec/cpg/enhancements/calls/ConstructorsTest.java
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.enhancements.calls;

import static org.junit.jupiter.api.Assertions.*;

import de.fraunhofer.aisec.cpg.BaseTest;
import de.fraunhofer.aisec.cpg.TestUtils;
import de.fraunhofer.aisec.cpg.graph.Node;
import de.fraunhofer.aisec.cpg.graph.declarations.ConstructorDeclaration;
import de.fraunhofer.aisec.cpg.graph.declarations.TranslationUnitDeclaration;
import de.fraunhofer.aisec.cpg.graph.declarations.VariableDeclaration;
import de.fraunhofer.aisec.cpg.graph.statements.expressions.*;
import java.nio.file.Path;
import java.util.List;
import org.junit.jupiter.api.Test;

class ConstructorsTest extends BaseTest {

  private final Path topLevel = Path.of("src", "test", "resources", "constructors");

  @Test
  void testJava() throws Exception {
    List<TranslationUnitDeclaration> result = TestUtils.analyze("java", topLevel, true);
    List<ConstructorDeclaration> constructors =
        TestUtils.subnodesOfType(result, ConstructorDeclaration.class);
    ConstructorDeclaration noArg =
        TestUtils.findByUniquePredicate(constructors, c -> c.getParameters().size() == 0);
    ConstructorDeclaration singleArg =
        TestUtils.findByUniquePredicate(constructors, c -> c.getParameters().size() == 1);
    ConstructorDeclaration twoArgs =
        TestUtils.findByUniquePredicate(constructors, c -> c.getParameters().size() == 2);

    List<VariableDeclaration> variables =
        TestUtils.subnodesOfType(result, VariableDeclaration.class);

    VariableDeclaration a1 = TestUtils.findByUniqueName(variables, "a1");
    assertTrue(a1.getInitializer() instanceof NewExpression);
    assertTrue(
        ((NewExpression) a1.getInitializer()).getInitializer() instanceof ConstructExpression);
    ConstructExpression a1Initializer =
        (ConstructExpression) ((NewExpression) a1.getInitializer()).getInitializer();
    assertEquals(noArg, a1Initializer.getConstructor());

    VariableDeclaration a2 = TestUtils.findByUniqueName(variables, "a2");
    assertTrue(a2.getInitializer() instanceof NewExpression);
    assertTrue(
        ((NewExpression) a2.getInitializer()).getInitializer() instanceof ConstructExpression);
    ConstructExpression a2Initializer =
        (ConstructExpression) ((NewExpression) a2.getInitializer()).getInitializer();
    assertEquals(singleArg, a2Initializer.getConstructor());

    VariableDeclaration a3 = TestUtils.findByUniqueName(variables, "a3");
    assertTrue(a3.getInitializer() instanceof NewExpression);
    assertTrue(
        ((NewExpression) a3.getInitializer()).getInitializer() instanceof ConstructExpression);
    ConstructExpression a3Initializer =
        (ConstructExpression) ((NewExpression) a3.getInitializer()).getInitializer();
    assertEquals(twoArgs, a3Initializer.getConstructor());

    VariableDeclaration a4 = TestUtils.findByUniqueName(variables, "a4");
    assertTrue(a4.getInitializer() instanceof UninitializedValue);
  }

  @Test
  void testCPP() throws Exception {
    List<TranslationUnitDeclaration> result = TestUtils.analyze("cpp", topLevel, true);
    List<ConstructorDeclaration> constructors =
        TestUtils.subnodesOfType(result, ConstructorDeclaration.class);
    ConstructorDeclaration noArg =
        TestUtils.findByUniquePredicate(
            constructors, c -> c.getParameters().size() == 0 && c.getName().equals("A"));
    ConstructorDeclaration singleArg =
        TestUtils.findByUniquePredicate(
            constructors, c -> c.getParameters().size() == 1 && c.getName().equals("A"));
    ConstructorDeclaration twoArgs =
        TestUtils.findByUniquePredicate(
            constructors, c -> c.getParameters().size() == 2 && c.getName().equals("A"));

    List<VariableDeclaration> variables =
        TestUtils.subnodesOfType(result, VariableDeclaration.class);

    VariableDeclaration a1 = TestUtils.findByUniqueName(variables, "a1");
    assertTrue(a1.getInitializer() instanceof ConstructExpression);
    ConstructExpression a1Initializer = (ConstructExpression) a1.getInitializer();
    assertEquals(noArg, a1Initializer.getConstructor());

    VariableDeclaration a2 = TestUtils.findByUniqueName(variables, "a2");
    assertTrue(a2.getInitializer() instanceof ConstructExpression);
    ConstructExpression a2Initializer = (ConstructExpression) a2.getInitializer();
    assertEquals(singleArg, a2Initializer.getConstructor());

    VariableDeclaration a3 = TestUtils.findByUniqueName(variables, "a3");
    assertTrue(a3.getInitializer() instanceof ConstructExpression);
    ConstructExpression a3Initializer = (ConstructExpression) a3.getInitializer();
    assertEquals(twoArgs, a3Initializer.getConstructor());

    VariableDeclaration a4 = TestUtils.findByUniqueName(variables, "a4");
    assertTrue(a4.getInitializer() instanceof ConstructExpression);
    ConstructExpression a4Initializer = (ConstructExpression) a4.getInitializer();
    assertEquals(noArg, a4Initializer.getConstructor());

    VariableDeclaration a5 = TestUtils.findByUniqueName(variables, "a5");
    assertTrue(a5.getInitializer() instanceof ConstructExpression);
    ConstructExpression a5Initializer = (ConstructExpression) a5.getInitializer();
    assertEquals(singleArg, a5Initializer.getConstructor());

    VariableDeclaration a6 = TestUtils.findByUniqueName(variables, "a6");
    assertTrue(a6.getInitializer() instanceof ConstructExpression);
    ConstructExpression a6Initializer = (ConstructExpression) a6.getInitializer();
    assertEquals(twoArgs, a6Initializer.getConstructor());

    VariableDeclaration a7 = TestUtils.findByUniqueName(variables, "a7");
    assertTrue(a7.getInitializer() instanceof NewExpression);
    assertTrue(
        ((NewExpression) a7.getInitializer()).getInitializer() instanceof ConstructExpression);
    ConstructExpression a7Initializer =
        (ConstructExpression) ((NewExpression) a7.getInitializer()).getInitializer();
    assertEquals(noArg, a7Initializer.getConstructor());

    VariableDeclaration a8 = TestUtils.findByUniqueName(variables, "a8");
    assertTrue(a8.getInitializer() instanceof NewExpression);
    assertTrue(
        ((NewExpression) a8.getInitializer()).getInitializer() instanceof ConstructExpression);
    ConstructExpression a8Initializer =
        (ConstructExpression) ((NewExpression) a8.getInitializer()).getInitializer();
    assertEquals(noArg, a8Initializer.getConstructor());

    VariableDeclaration a9 = TestUtils.findByUniqueName(variables, "a9");
    assertTrue(a9.getInitializer() instanceof NewExpression);
    assertTrue(
        ((NewExpression) a9.getInitializer()).getInitializer() instanceof ConstructExpression);
    ConstructExpression a9Initializer =
        (ConstructExpression) ((NewExpression) a9.getInitializer()).getInitializer();
    assertEquals(singleArg, a9Initializer.getConstructor());

    VariableDeclaration a10 = TestUtils.findByUniqueName(variables, "a10");
    assertTrue(a10.getInitializer() instanceof NewExpression);
    assertTrue(
        ((NewExpression) a10.getInitializer()).getInitializer() instanceof ConstructExpression);
    ConstructExpression a10Initializer =
        (ConstructExpression) ((NewExpression) a10.getInitializer()).getInitializer();
    assertEquals(twoArgs, a10Initializer.getConstructor());
  }

  @Test
  void testCPPFullDefault() throws Exception {
    List<TranslationUnitDeclaration> result =
        TestUtils.analyze(
            List.of(Path.of(topLevel.toString(), "defaultarg", "constructorDefault.cpp").toFile()),
            topLevel,
            true);
    List<ConstructorDeclaration> constructors =
        TestUtils.subnodesOfType(result, ConstructorDeclaration.class);

    List<VariableDeclaration> variables =
        TestUtils.subnodesOfType(result, VariableDeclaration.class);

    ConstructorDeclaration twoDefaultArg =
        TestUtils.findByUniquePredicate(
            constructors, c -> c.getDefaultParameters().size() == 2 && c.getName().equals("D"));

    Literal<?> literal0 =
        TestUtils.findByUniquePredicate(
            TestUtils.subnodesOfType(result, Literal.class), l -> l.getValue().equals(0));
    Literal<?> literal1 =
        TestUtils.findByUniquePredicate(
            TestUtils.subnodesOfType(result, Literal.class), l -> l.getValue().equals(1));

    VariableDeclaration d1 = TestUtils.findByUniqueName(variables, "d1");
    assertTrue(d1.getInitializer() instanceof ConstructExpression);
    ConstructExpression d1Initializer = (ConstructExpression) d1.getInitializer();
    assertEquals(twoDefaultArg, d1Initializer.getConstructor());
    assertEquals(0, d1Initializer.getArguments().size());
    assertTrue(twoDefaultArg.getNextEOG().contains(literal0));
    assertTrue(twoDefaultArg.getNextEOG().contains(literal1));
    assertTrue(literal0.getNextEOG().contains(literal1));
    for (Node node : twoDefaultArg.getNextEOG()) {
      if (!(node.equals(literal0) || node.equals(literal1))) {
        assertTrue(literal1.getNextEOG().contains(node));
      }
    }

    VariableDeclaration d2 = TestUtils.findByUniqueName(variables, "d2");
    assertTrue(d2.getInitializer() instanceof ConstructExpression);
    ConstructExpression d2Initializer = (ConstructExpression) d2.getInitializer();
    assertEquals(twoDefaultArg, d2Initializer.getConstructor());
    assertEquals(1, d2Initializer.getArguments().size());
    assertEquals(2, ((Literal) d2Initializer.getArguments().get(0)).getValue());

    VariableDeclaration d3 = TestUtils.findByUniqueName(variables, "d3");
    assertTrue(d3.getInitializer() instanceof ConstructExpression);
    ConstructExpression d3Initializer = (ConstructExpression) d3.getInitializer();
    assertEquals(twoDefaultArg, d3Initializer.getConstructor());
    assertEquals(2, d3Initializer.getArguments().size());
    assertEquals(3, ((Literal) d3Initializer.getArguments().get(0)).getValue());
    assertEquals(4, ((Literal) d3Initializer.getArguments().get(1)).getValue());
  }

  @Test
  void testCPPPartialDefault() throws Exception {
    List<TranslationUnitDeclaration> result =
        TestUtils.analyze(
            List.of(Path.of(topLevel.toString(), "defaultarg", "constructorDefault.cpp").toFile()),
            topLevel,
            true);
    List<ConstructorDeclaration> constructors =
        TestUtils.subnodesOfType(result, ConstructorDeclaration.class);

    List<VariableDeclaration> variables =
        TestUtils.subnodesOfType(result, VariableDeclaration.class);

    ConstructorDeclaration singleDefaultArg =
        TestUtils.findByUniquePredicate(
            constructors, c -> c.getParameters().size() == 2 && c.getName().equals("E"));

    Literal<?> literal10 =
        TestUtils.findByUniquePredicate(
            TestUtils.subnodesOfType(result, Literal.class), l -> l.getValue().equals(10));

    VariableDeclaration e1 = TestUtils.findByUniqueName(variables, "e1");
    assertTrue(e1.getInitializer() instanceof ConstructExpression);
    ConstructExpression e1Initializer = (ConstructExpression) e1.getInitializer();
    assertTrue(e1Initializer.getConstructor().isImplicit());
    assertEquals(0, e1Initializer.getArguments().size());

    VariableDeclaration e2 = TestUtils.findByUniqueName(variables, "e2");
    assertTrue(e2.getInitializer() instanceof ConstructExpression);
    ConstructExpression e2Initializer = (ConstructExpression) e2.getInitializer();
    assertEquals(singleDefaultArg, e2Initializer.getConstructor());
    assertEquals(1, e2Initializer.getArguments().size());
    assertEquals(5, ((Literal) e2Initializer.getArguments().get(0)).getValue());
    assertTrue(singleDefaultArg.getNextEOG().contains(literal10));
    for (Node node : singleDefaultArg.getNextEOG()) {
      if (!node.equals(literal10)) {
        assertTrue(literal10.getNextEOG().contains(node));
      }
    }

    VariableDeclaration e3 = TestUtils.findByUniqueName(variables, "e3");
    assertTrue(e3.getInitializer() instanceof ConstructExpression);
    ConstructExpression e3Initializer = (ConstructExpression) e3.getInitializer();
    assertEquals(singleDefaultArg, e3Initializer.getConstructor());
    assertEquals(2, e3Initializer.getArguments().size());
    assertEquals(6, ((Literal) e3Initializer.getArguments().get(0)).getValue());
    assertEquals(7, ((Literal) e3Initializer.getArguments().get(1)).getValue());
  }

  @Test
  void testCPPImplicitCast() throws Exception {
    List<TranslationUnitDeclaration> result =
        TestUtils.analyze(
            List.of(
                Path.of(topLevel.toString(), "implicitcastarg", "constructorImplicit.cpp")
                    .toFile()),
            topLevel,
            true);
    List<ConstructorDeclaration> constructors =
        TestUtils.subnodesOfType(result, ConstructorDeclaration.class);

    List<VariableDeclaration> variables =
        TestUtils.subnodesOfType(result, VariableDeclaration.class);

    ConstructorDeclaration implicitConstructor =
        TestUtils.findByUniquePredicate(constructors, c -> c.getName().equals("I"));

    Literal<?> literal10 =
        TestUtils.findByUniquePredicate(
            TestUtils.subnodesOfType(result, Literal.class), l -> l.getValue().equals(10));

    VariableDeclaration i1 = TestUtils.findByUniqueName(variables, "i1");
    assertTrue(i1.getInitializer() instanceof ConstructExpression);
    ConstructExpression i1Constructor = (ConstructExpression) i1.getInitializer();
    assertFalse(i1Constructor.isImplicit());

    assertEquals(implicitConstructor, i1Constructor.getConstructor());
    assertEquals(1, i1Constructor.getArguments().size());
    assertTrue(i1Constructor.getArguments().get(0) instanceof CastExpression);
    CastExpression i1ConstructorArgument = (CastExpression) i1Constructor.getArguments().get(0);
    assertEquals("int", i1ConstructorArgument.getCastType().getName());
    assertEquals("1.0", i1ConstructorArgument.getExpression().getCode());
    assertEquals("double", i1ConstructorArgument.getExpression().getType().getName());

    ConstructorDeclaration implicitConstructorWithDefault =
        TestUtils.findByUniquePredicate(constructors, c -> c.getName().equals("H"));

    VariableDeclaration h1 = TestUtils.findByUniqueName(variables, "h1");
    assertTrue(h1.getInitializer() instanceof ConstructExpression);
    ConstructExpression h1Constructor = (ConstructExpression) h1.getInitializer();
    assertFalse(h1Constructor.isImplicit());

    assertEquals(implicitConstructorWithDefault, h1Constructor.getConstructor());
    assertEquals(1, h1Constructor.getArguments().size());

    assertTrue(h1Constructor.getArguments().get(0) instanceof CastExpression);
    CastExpression h1ConstructorArgument1 = (CastExpression) h1Constructor.getArguments().get(0);
    assertEquals("int", h1ConstructorArgument1.getCastType().getName());
    assertEquals("2.0", h1ConstructorArgument1.getExpression().getCode());
    assertEquals("double", h1ConstructorArgument1.getExpression().getType().getName());

    assertTrue(implicitConstructorWithDefault.getNextEOG().contains(literal10));
    for (Node node : implicitConstructorWithDefault.getNextEOG()) {
      if (!node.equals(literal10)) {
        assertTrue(literal10.getNextEOG().contains(node));
      }
    }
  }
}

File: cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/enhancements/calls/ConstructorsTest.kt
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.enhancements.calls

import de.fraunhofer.aisec.cpg.BaseTest
import de.fraunhofer.aisec.cpg.TestUtils
import de.fraunhofer.aisec.cpg.TestUtils.findByUniquePredicate
import de.fraunhofer.aisec.cpg.graph.*
import de.fraunhofer.aisec.cpg.graph.SearchModifier.UNIQUE
import de.fraunhofer.aisec.cpg.graph.allChildren
import de.fraunhofer.aisec.cpg.graph.declarations.ConstructorDeclaration
import de.fraunhofer.aisec.cpg.graph.statements.expressions.*
import de.fraunhofer.aisec.cpg.graph.variables
import java.nio.file.Path
import kotlin.test.*

internal class ConstructorsTest : BaseTest() {
    private val topLevel = Path.of("src", "test", "resources", "constructors")

    @Test
    @Throws(Exception::class)
    fun testJava() {
        val result = TestUtils.analyze("java", topLevel, true)
        val constructors = result.allChildren<ConstructorDeclaration>()
        val noArg = findByUniquePredicate(constructors) { it.parameters.size == 0 }
        val singleArg = findByUniquePredicate(constructors) { it.parameters.size == 1 }
        val twoArgs = findByUniquePredicate(constructors) { it.parameters.size == 2 }
        val variables = result.variables
        val a1 = assertNotNull(variables["a1", UNIQUE])
        assertNotNull(a1)
        assertTrue(a1.initializer is NewExpression)
        assertTrue((a1.initializer as? NewExpression)?.initializer is ConstructExpression)

        val a1Initializer = (a1.initializer as NewExpression).initializer as ConstructExpression
        assertEquals(noArg, a1Initializer.constructor)

        val a2 = assertNotNull(variables["a2", UNIQUE])
        assertNotNull(a2)
        assertTrue(a2.initializer is NewExpression)
        assertTrue((a2.initializer as? NewExpression)?.initializer is ConstructExpression)

        val a2Initializer = (a2.initializer as NewExpression).initializer as ConstructExpression
        assertEquals(singleArg, a2Initializer.constructor)

        val a3 = assertNotNull(variables["a3", UNIQUE])
        assertNotNull(a3)
        assertTrue(a3.initializer is NewExpression)
        assertTrue((a3.initializer as NewExpression).initializer is ConstructExpression)

        val a3Initializer = (a3.initializer as NewExpression).initializer as ConstructExpression
        assertEquals(twoArgs, a3Initializer.constructor)

        val a4 = assertNotNull(variables["a4", UNIQUE])
        assertNotNull(a4)
        assertTrue(a4.initializer is UninitializedValue)
    }

    @Test
    @Throws(Exception::class)
    fun testCPP() {
        val result = TestUtils.analyze("cpp", topLevel, true)
        val constructors = result.allChildren<ConstructorDeclaration>()
        val noArg =
            findByUniquePredicate(constructors) { it.parameters.size == 0 && it.name == "A" }
        val singleArg =
            findByUniquePredicate(constructors) { it.parameters.size == 1 && it.name == "A" }
        val twoArgs =
            findByUniquePredicate(constructors) { it.parameters.size == 2 && it.name == "A" }
        val variables = result.variables
        val a1 = assertNotNull(variables["a1", UNIQUE])
        assertNotNull(a1)
        assertTrue(a1.initializer is ConstructExpression)

        val a1Initializer = a1.initializer as ConstructExpression
        assertEquals(noArg, a1Initializer.constructor)

        val a2 = assertNotNull(variables["a2", UNIQUE])
        assertNotNull(a2)
        assertTrue(a2.initializer is ConstructExpression)

        val a2Initializer = a2.initializer as ConstructExpression
        assertEquals(singleArg, a2Initializer.constructor)

        val a3 = assertNotNull(variables["a3", UNIQUE])
        assertNotNull(a3)
        assertTrue(a3.initializer is ConstructExpression)

        val a3Initializer = a3.initializer as ConstructExpression
        assertEquals(twoArgs, a3Initializer.constructor)

        val a4 = assertNotNull(variables["a4", UNIQUE])
        assertNotNull(a4)
        assertTrue(a4.initializer is ConstructExpression)

        val a4Initializer = a4.initializer as ConstructExpression
        assertEquals(noArg, a4Initializer.constructor)

        val a5 = assertNotNull(variables["a5", UNIQUE])
        assertNotNull(a5)
        assertTrue(a5.initializer is ConstructExpression)

        val a5Initializer = a5.initializer as ConstructExpression
        assertEquals(singleArg, a5Initializer.constructor)

        val a6 = assertNotNull(variables["a6", UNIQUE])
        assertNotNull(a6)
        assertTrue(a6.initializer is ConstructExpression)

        val a6Initializer = a6.initializer as ConstructExpression
        assertEquals(twoArgs, a6Initializer.constructor)

        val a7 = assertNotNull(variables["a7", UNIQUE])
        assertNotNull(a7)
        assertTrue(a7.initializer is NewExpression)
        assertTrue((a7.initializer as NewExpression).initializer is ConstructExpression)

        val a7Initializer = (a7.initializer as NewExpression).initializer as ConstructExpression
        assertEquals(noArg, a7Initializer.constructor)

        val a8 = assertNotNull(variables["a8", UNIQUE])
        assertNotNull(a8)
        assertTrue(a8.initializer is NewExpression)
        assertTrue((a8.initializer as NewExpression).initializer is ConstructExpression)

        val a8Initializer = (a8.initializer as NewExpression).initializer as ConstructExpression
        assertEquals(noArg, a8Initializer.constructor)

        val a9 = assertNotNull(variables["a9", UNIQUE])
        assertNotNull(a9)
        assertTrue(a9.initializer is NewExpression)
        assertTrue((a9.initializer as NewExpression).initializer is ConstructExpression)

        val a9Initializer = (a9.initializer as NewExpression).initializer as ConstructExpression
        assertEquals(singleArg, a9Initializer.constructor)

        val a10 = assertNotNull(variables["a10", UNIQUE])
        assertNotNull(a10)
        assertTrue(a10.initializer is NewExpression)
        assertTrue((a10.initializer as NewExpression).initializer is ConstructExpression)

        val a10Initializer = (a10.initializer as NewExpression).initializer as ConstructExpression
        assertEquals(twoArgs, a10Initializer.constructor)
    }

    @Test
    @Throws(Exception::class)
    fun testCPPFullDefault() {
        val result =
            TestUtils.analyze(
                listOf(
                    Path.of(topLevel.toString(), "defaultarg", "constructorDefault.cpp").toFile()
                ),
                topLevel,
                true
            )
        val constructors = result.allChildren<ConstructorDeclaration>()
        val variables = result.variables
        val twoDefaultArg =
            findByUniquePredicate(constructors) { it.defaultParameters.size == 2 && it.name == "D" }
        assertNotNull(twoDefaultArg)

        val literal0 = findByUniquePredicate(result.literals) { it.value == 0 }
        val literal1 = findByUniquePredicate(result.literals) { it.value == 1 }
        val d1 = assertNotNull(variables["d1", UNIQUE])
        assertNotNull(d1)
        assertTrue(d1.initializer is ConstructExpression)

        val d1Initializer = d1.initializer as ConstructExpression
        assertEquals(twoDefaultArg, d1Initializer.constructor)
        assertEquals(0, d1Initializer.arguments.size)
        assertTrue(twoDefaultArg.nextEOG.contains(literal0))
        assertTrue(twoDefaultArg.nextEOG.contains(literal1))
        assertTrue(literal0.nextEOG.contains(literal1))

        for (node in twoDefaultArg.nextEOG) {
            if (!(node == literal0 || node == literal1)) {
                assertTrue(literal1.nextEOG.contains(node))
            }
        }
        val d2 = assertNotNull(variables["d2", UNIQUE])
        assertTrue(d2.initializer is ConstructExpression)

        val d2Initializer = d2.initializer as ConstructExpression
        assertEquals(twoDefaultArg, d2Initializer.constructor)
        assertEquals(1, d2Initializer.arguments.size)
        assertEquals(2, (d2Initializer.arguments[0] as Literal<*>).value)

        val d3 = assertNotNull(variables["d3", UNIQUE])
        assertTrue(d3.initializer is ConstructExpression)

        val d3Initializer = d3.initializer as ConstructExpression
        assertEquals(twoDefaultArg, d3Initializer.constructor)
        assertEquals(2, d3Initializer.arguments.size)
        assertEquals(3, (d3Initializer.arguments[0] as Literal<*>).value)
        assertEquals(4, (d3Initializer.arguments[1] as Literal<*>).value)
    }

    @Test
    @Throws(Exception::class)
    fun testCPPPartialDefault() {
        val result =
            TestUtils.analyze(
                listOf(
                    Path.of(topLevel.toString(), "defaultarg", "constructorDefault.cpp").toFile()
                ),
                topLevel,
                true
            )
        val constructors = result.allChildren<ConstructorDeclaration>()
        val variables = result.variables
        val singleDefaultArg =
            findByUniquePredicate(constructors) { c: ConstructorDeclaration ->
                c.parameters.size == 2 && c.name == "E"
            }
        val literal10 = findByUniquePredicate(result.literals) { it.value == 10 }
        val e1 = assertNotNull(variables["e1", UNIQUE])
        assertTrue(e1.initializer is ConstructExpression)
        val e1Initializer = e1.initializer as ConstructExpression
        assertTrue(e1Initializer.constructor!!.isInferred)
        assertEquals(0, e1Initializer.arguments.size)
        val e2 = assertNotNull(variables["e2", UNIQUE])
        assertTrue(e2.initializer is ConstructExpression)
        val e2Initializer = e2.initializer as ConstructExpression
        assertEquals(singleDefaultArg, e2Initializer.constructor)
        assertEquals(1, e2Initializer.arguments.size)
        assertEquals(5, (e2Initializer.arguments[0] as Literal<*>).value)
        assertTrue(singleDefaultArg.nextEOG.contains(literal10))
        for (node in singleDefaultArg.nextEOG) {
            if (node != literal10) {
                assertTrue(literal10.nextEOG.contains(node))
            }
        }
        val e3 = assertNotNull(variables["e3", UNIQUE])
        assertTrue(e3.initializer is ConstructExpression)
        val e3Initializer = e3.initializer as ConstructExpression
        assertEquals(singleDefaultArg, e3Initializer.constructor)
        assertEquals(2, e3Initializer.arguments.size)
        assertEquals(6, (e3Initializer.arguments[0] as Literal<*>).value)
        assertEquals(7, (e3Initializer.arguments[1] as Literal<*>).value)
    }

    @Test
    @Throws(Exception::class)
    fun testCPPImplicitCast() {
        val result =
            TestUtils.analyze(
                listOf(
                    Path.of(topLevel.toString(), "implicitcastarg", "constructorImplicit.cpp")
                        .toFile()
                ),
                topLevel,
                true
            )
        val constructors = result.allChildren<ConstructorDeclaration>()
        val variables = result.variables
        val implicitConstructor =
            findByUniquePredicate(constructors) { c: ConstructorDeclaration -> c.name == "I" }
        val literal10 = findByUniquePredicate(result.literals) { it.value == 10 }
        val i1 = assertNotNull(variables["i1", UNIQUE])
        assertTrue(i1.initializer is ConstructExpression)

        val i1Constructor = i1.initializer as ConstructExpression
        assertFalse(i1Constructor.isImplicit)
        assertEquals(implicitConstructor, i1Constructor.constructor)
        assertEquals(1, i1Constructor.arguments.size)
        assertTrue(i1Constructor.arguments[0] is CastExpression)

        val i1ConstructorArgument = i1Constructor.arguments[0] as CastExpression
        assertEquals("int", i1ConstructorArgument.castType.name)
        assertEquals("1.0", i1ConstructorArgument.expression.code)
        assertEquals("double", i1ConstructorArgument.expression.type.name)

        val implicitConstructorWithDefault =
            findByUniquePredicate(constructors) { c: ConstructorDeclaration -> c.name == "H" }
        val h1 = assertNotNull(variables["h1", UNIQUE])
        assertTrue(h1.initializer is ConstructExpression)

        val h1Constructor = h1.initializer as ConstructExpression
        assertFalse(h1Constructor.isImplicit)
        assertEquals(implicitConstructorWithDefault, h1Constructor.constructor)
        assertEquals(1, h1Constructor.arguments.size)
        assertTrue(h1Constructor.arguments[0] is CastExpression)

        val h1ConstructorArgument1 = h1Constructor.arguments[0] as CastExpression
        assertEquals("int", h1ConstructorArgument1.castType.name)
        assertEquals("2.0", h1ConstructorArgument1.expression.code)
        assertEquals("double", h1ConstructorArgument1.expression.type.name)
        assertTrue(implicitConstructorWithDefault.nextEOG.contains(literal10))
        for (node in implicitConstructorWithDefault.nextEOG) {
            if (node != literal10) {
                assertTrue(literal10.nextEOG.contains(node))
            }
        }
    }
}
